com.unity.netcode.gameobjects/Components/NetworkAnimator.cs

```diff
@@ -1,79 +1,167 @@
             }
 
             public AnimatorSnapshot()
             {
-                BoolParameters = new Dictionary<int, bool>(0);
-                FloatParameters = new Dictionary<int, float>(0);
-                IntParameters = new Dictionary<int, int>(0);
-                TriggerParameters = new HashSet<int>();
+                BoolParamArray = new KeyValuePair<int, bool>[0];
+                IntParamArray = new KeyValuePair<int, int>[0];
+                FloatParamArray = new KeyValuePair<int, float>[0];
+                TriggerParameters = new List<int>(0);
                 LayerStates = new LayerState[0];
+                SetBuffersToDefaultValues();
+            }
+
+            private void SetBuffersToDefaultValues()
+            {
+                for (int i = 0; i < BoolParamArray.Length; i++)
+                {
+                    BoolParamArray[i] = new KeyValuePair<int, bool>(k_InvalidKey, false);
+                }
+
+                for (int i = 0; i < FloatParamArray.Length; i++)
+                {
+                    FloatParamArray[i] = new KeyValuePair<int, float>(k_InvalidKey, 0);
+                }
+
+                for (int i = 0; i < IntParamArray.Length; i++)
+                {
+                    IntParamArray[i] = new KeyValuePair<int, int>(k_InvalidKey, 0);
+                }
+
+                TriggerParameters.Clear();
+
+                Array.Clear(LayerStates, 0, LayerStates.Length);
             }
 
             public bool SetInt(int key, int value)
             {
-                if (IntParameters.TryGetValue(key, out var existingValue) && existingValue == value)
+                bool setOrUpdatedValue = false;
+
+                int existingKvIndex = Array.FindIndex(IntParamArray, pair => pair.Key == key);
+
+                if (existingKvIndex == -1)
                 {
-                    return false;
+                    for (int i = 0; i < IntParamArray.Length; i++)
+                    {
+                        var kv = IntParamArray[i];
+
+                        if (kv.Key == k_InvalidKey)
+                        {
+                            IntParamArray[i] = new KeyValuePair<int, int>(key, value);
+                            setOrUpdatedValue = true;
+                            break;
+                        }
+                    }
+                }
+                else if (IntParamArray[existingKvIndex].Value != value)
+                {
+                    IntParamArray[existingKvIndex] = new KeyValuePair<int, int>(key, value);
+                    setOrUpdatedValue = true;
                 }
 
-                IntParameters[key] = value;
-                return true;
+                return setOrUpdatedValue;
             }
 
             public bool SetBool(int key, bool value)
             {
-                if (BoolParameters.TryGetValue(key, out var existingValue) && existingValue == value)
+                bool setOrUpdatedValue = false;
```

Comment:
It's too bad there isn't a clever way to merge SetBool / SetInt / SetFloat into one function but I see the problem of dealing with the KeyValuePair's of different contained types